man/ffptxt.1.in
.de CW
. nop \s-2\f[C]\\$*\f[]\s+2
..
.de CODE
.in +0.5i
. nop \s-2\f[C]\\$*\f[R]\s+2
.in -0.5i
..
.TH ffptxt 1 "[@]DATE[@]" "Version [@]VERSION[@]" "FFP PHYLOGENY"
.SH NAME
ffptxt \- This program generates an FFP vector of text data.
.SH SYNOPSIS
.BI "ffpaa [" "OPTION" "] ... [" "FILE"  "] ..."
.SH DESCRIPTION
.PP
Given no options at all the default behavior of the program is to
generate a Feature Frequency Profile (FFP) of text data, using 26
alphabet letters, removing all non-alphabetic characters.  The 
default feature length is 6.
Text will be read from standard input if no options are supplied and 
.B ffptxt
is called non-interactively (i.e. as part of a pipeline) or with a "-" in the 
argument list.
.
.SH OPTIONS
.TP
.BI "\-l " "LEN" ", --length=" "LEN"
.RI "Changes the default length of features to " "LEN" ". The default length is 6."
The maximum word size is 40.
You may overide this maximum length by setting (and exporting)
the environmental variable MAX_WORD_SIZE to a different value.
.TP
.BI "\-f " "FILE" ", --feature-list=" "FILE"
Changes the behavior of the program to read
.RI "a list of features from " "FILE" ". Features can"
be space or newline delimited.
.TP
.B "\-h, --help"
Display help message.
.PP
.SH EXAMPLES
The command:
.PP
.CODE ffptxt -l 3 file*.txt
.PP
Will produce a output of this form, using features of length 3:
.PP
.CODE hel 2 elo 3 low 4 ...
.CODE thi 3 ism 2 sme 5 ...
.PP
There will be one FFP line for each .txt file
.PP
This format can also be used:
.PP
.CODE ffptxt -l 4 file1.txt file2.txt file3.txt
.PP
The utilities in the FFP package can be used as fully
qualified filters and can be used to build pipelines.
This pipeline will create a phylip infile format distance
matrix, using the Jensen Shannon Divergence:
.PP
.CODE ffptxt -l 7 file*.txt | ffpcol -t | ffprwn | ffpjsd -p > infile
.PP
.RB "Please note the use of the" " -t " "option in ffpcol to specify text."
.PP
Bootstrap and jacknife resampling can be done with the
ffpboot program.
.PP
.SH FUTURE DIRECTIONS
None.
.PP
.SH AUTHOR
This program was written by Gregory E. Sims.
.SH "REPORTING BUGS"
Report bugs to <gesims@lbl.gov>.
.SH COPYRIGHT
Copyright (C) [@]COPY[@] Gregory E. Sims
.br
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR ffpry(1),
.BR ffprwn(1), 
.BR ffpjsd(1),
.BR ffpcol(1),
.BR ffpmerge(1),
.BR ffpboot(1),
.BR ffpvprof(1),
.BR ffpreprof(1)


NEWS
Version 3.19
	Thanks to Joemar Taggana, some incompatible utilities 
	were discovered in Mac OSX -- including incompatible
	getopt and mktemp.  Corrected confusion in tutorial 
        p. 3 regarding -m (Thanks to Erin Shellman).  Corrected
	spelling of ffpjsd option --hamann to --hamman.

Version 3.18
	Rearranged ungetc character tests to come after
	conversion of pipe to file.  Fixed memory leak in
	hash init().  Added support for large files (> 2GB) on 32
        bit machines.

Version 3.17
	Fixed bug in tempfile creation in function 
        convertPipeToFile().

Version 3.16
	Fixed bug in ffpry with multiple headers parsed as a
	single genome entry.

Version 3.15
	Fixed bug in ffprwn.

Version 3.14
	Added --disable-gui options to ./configure script. This
	disables the perl based gui, so that the Perl/Tk module
	isn't required for installation.

Version 3.13
	Changed row terminal output for all programs from '\t\n' 
	(tab newline) to '\n' (newline) only.  This makes output
	from previous version incompatible with 3.13 but the
	ffps can be easily converted using:

	sed 's/\t$//' ffp > ffp.tmp ; mv ffp.tmp ffp

Version 3.12
	Small bug fix for ffprwn for GNU compiler on 64 bit machines
	where promotion of unsigned int to long unsigned int is
	different.  Thanks to Adrienne Breland at the University
	of Nevada, Reno for finding this.

Version 3.10
	Fixed minor bug in ffpry for fasta records pseudomolecule 
	input ENDING in NNNNNNNN, or other invalid characters.

Version 3.09
	Fixed bug in ffpdf, related to initialized GROUP variable.
	
	Updated license

Version 3.08

	ffpgui

	Added instructions to install a functional version of the
	perl/tk module for using cygwin, in INSTALL.

	Added error messages in ffpgui for running a vocab or 
	relative entropy profile without loading a file first.

	Added additional sections to tutorial.pdf
       


Version 3.07
	Added signal handling to all executables.  
	Broken pipes will cause a termination error in a pipeline
	and will output 'PROGRAM: Broken pipe'

	Expanded ./doc/tutorial.in.
	Updated license.
	Fixed inappropriate default -f setting (changed from 1 to
	2) for ffpvprof.

Version 3.06
	Added ffptree utility, which creates neighbor joining and
	upgma trees.  Date and version automatically updated 
	during build process.
	
        Improved seeding of random generator when no seed is
	provide via command line option:  ffpry, ffpboot, ffpaa

	Added processing via ffptree to ffpgui.  The human 
	readable tree is displayed in the message window.

	Started tutorial file in ./doc directory. 

	Added tutorial.pdf to doc directory

Version 3.05
	Added length test since MAX_WORD_SIZE is 40: ffpry, ffpaa
	ffpre.

Version 3.04
	Added script ffpgui which is a Perl/Tk based GUI wrapper of
	the various utilities.

	Added --keys flag to ffpfilt to force keys to be output
	instead of values only.

Version 3.03
	Fixed overflow error with non-rabin karp hash used in ffpcol.

Version 3.02
	The ffpaat utility has been removed from this package.  The
	purpose of this utility was to count amino acid features in 
	all nine frames from DNA fasta input.  However	obtaining
	a proper translation of gene regions and using ffpaa is 
	generally more informative.  This utility can still be 
	compiled and used from previous versions, but will not longer
	be part of future releases. 

	Updated ffpre to Rabin-Karp hash, also fixed float overflow
	errors and cases of nan, or inf in output.  Text mode is now
	implemented in ffpre.

	Updated ffptxt to Rabin-Karp hash.


Version 3.01
	Implementation of Rabin-Karp hash in ffpaa utility.
	Similar speed gains as in ffpry obtained


	ffpaa -l 20 NC_008253.faa      

	ffpaa (v.3.0)                        ----------- 9.35 s 
	ffpaa (v.3.01)                       ----------- 0.15 s

	Input buffering options removed from both ffpaa and ffpry. The
	optimal buffering size for the system is determined automatically
	by using the stat.st_block size attribute of fstat or the compiler
	default, whichever is greater.  This is an effort to use the
	filesystem's most efficient block size.  
	
	Upcoming v3.02 will include Rabin-Karp implementation of ffpre.

Version 3.0
	Thanks to Nick Loman of University of Birmingham for pointing out
	the cryptic nature of the "key-value" error messages when both
	ffpfilt and ffpcol are used together in a pipeline successively.
	
	All error messages from the utilities state the program issuing
	the error.

	Revamped tmpfile creation for non-seekable piped input, to use
	most efficient buffering.

	Added extra unit tests to confirm piped vs file argument output
	is identical.

	ffpry changes:

	Beginning initial development of Rabin Karp Rolling hash function
	implementations which has replaced the hashing function used
	in version 2 in the ffpry executable.  This has provided a 
	significant speedup in k-mer counting.
	Here is a speed comparison:
	
	ffpry -l 10 -d NC_008253.fna             ------    9.7s (v.3.0)
        ffpry -l 10 -d NC_008253.fna             ------ 1m28.2s (v.2.30

	Removed all-possible/-a flag from ffpry.  The only way to generate
	columnar output from ffpry is via the -f flag and specify a specific
	feature set.

        -f specified features are now reverse complemented and classed as
	is the input sequence by default.	

	Added -r option which disables reverse complementing of the input
	sequence as well as operates on the feature list.  
	
	Fixed ffpry -f long option to correct value as per man file.
	
	Other:

	Fixed segmentation error in ffpboot

	Future directions:

	Rabin-Karp hash function will be implemented in ffpaa, ffptxt and
	ffpre, ffpaat
	
	

Version 2.30
	Fixed bug when encountering IUPAC codes in input files to ffpry.

Version 2.29
	Fixed bug when piping input into ffpmerge.

Version 2.28
	Addition of examples directory with trees from 
	Sims & Kim(2011) PNAS,108,8329-8334

        BUGS: ffpre doesn't report errors for trying to open

src/ffptxt.c
/*****************************************************
* This code is distributed under a Non-commercial use 
* license.  For details see LICENSE.  Use of this
* code must be properly attributed to its author
* Gregory E. Sims provided that its use or derivative 
* use is non-commercial in nature.  Proper attribution        
* can be made by citing:
*
* Sims GE, et al (2009) Alignment-free genome 
* comparison with feature frequency profiles (FFP) and 
* optimal resolutions. Proc. Natl. Acad. Sci. USA.
* 106, 2677-82.
*
* Gregory E. Sims (C) 2010-2012
*
*****************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <getopt.h>
#include <math.h>
#include <ctype.h>
#include <limits.h>
#include <errno.h>
#include <unistd.h>
#include <sys/stat.h>
#include "hashroll.h"
#include "mask.h"
#include "utils.h"
#include "vstring.h"
#include "sighandle.h"
#include "parse_features.h"
#include "../config.h"

/** @todo Need to implement stop word removal i.e. 'the' 'a' 'and' */

char PROG_NAME[FILENAME_MAX];
#define DEFAULT_WORD_LENGTH 6 /**< Default Feature length if not given by opt -l */

void parseFile(HASH *, FILE *);
void loopFeatureList(HASH * h, FILE * fp);
void loopRaw(HASH * h, FILE * fp);

char usage_str[] = "Usage: %s [OPTION] ... [FILE] ...\n\
This program generates an FFP vector from text data\n\n\
Given no options the default behavior of the program is to\n\
generate a Feature Frequency Profile (FFP) using features of\n\
length\n\
\t-l LEN, --length\n\
\t-f FILE, --feature-list\n\
\t-v, --version\n\
\t-h, --help\n\n\
Copyright (c) %s\n\
%s\n\
Contact %s\n";


int Length = DEFAULT_WORD_LENGTH;
				/**< Feature length to use if not specified by opt -l */
int maxWordSize = MAX_WORD_SIZE;
char *weightVector;



char *fvalue = NULL;  /**< -f File name to load features from */
char fflag = 0;	      /**< -f Option to only count features listed in a file */
char mflag = 0;       /**< unused but needed to link with hashroll.c */
bool zflag = false;
bool wflag = false;

int main(int argc, char **argv)
{
    int opt;
    HASH h;
    FILE *fp;

    int option_index = 0;

    static struct option long_options[] = {
	{"length", required_argument, 0, 'l'},
	{"feature-list", required_argument, 0, 'f'},
	{"help", no_argument, 0, 'h'},
	{"version", no_argument, 0, 'v'},
	{0, 0, 0, 0}
    };

    initSignalHandlers();

    strcpy(PROG_NAME,basename( argv[0] ));

    while ((opt = getopt_long(argc, argv, "l:f:hv",
			      long_options, &option_index)) != -1)

	switch (opt) {
	case 'l':
	    Length = atoi(optarg);
	    break;
	case 'f':
	    fflag = 1;
	    fvalue = optarg;
	    break;
	case 'v':
	    printVersion();
	    exit(EXIT_SUCCESS);
	    break;
	case 'h':
	    printUsageStr();
	    exit(EXIT_SUCCESS);
	    break;
	default:
	    printErrorUsageStr();
	    break;
	}

    if ( getenv("MAX_WORD_SIZE") )
	maxWordSize=atoi(getenv("MAX_WORD_SIZE"));	    

    if (Length > maxWordSize) 
	    fatal_msg("%d: Max Word size is : %d",Length,maxWordSize);


    init(&h, 0, text, 0, 0, Length);


    // If provided a feature list read it and store in hash

    if (fflag) 
	    parseFeatureList(&h,fvalue,Length,text);
    

// Must now process file arguments
    argv += optind;

    do {
	fp = stdin;
	if (*argv) {
	    if (!strcmp(*argv, "-"))
		fp = stdin;
	    else if ((fp = fopen(*argv, "r")) == NULL)
		fatal_msg("%s: %s.\n", *argv,strerror(errno));

	    if ( isDirectory(*argv) ) 
	    	fatal_msg("%s: %s\n",*argv, strerror(EISDIR));

	    argv++;
	} else if (isatty(STDIN_FILENO))
	    printErrorUsageStr();

	parseFile(&h, fp);

	if (fp != stdin)
	    fclose(fp);

    } while (*argv);

    freeHash(&h);

    return EXIT_SUCCESS;
}



void parseFile(HASH * h, FILE * fp)
{

// to eliminate a test w/n a tight loop
// First test here.
    if (fflag)
	loopFeatureList(h, fp);
    else
	loopRaw(h, fp);

}


// For finding only features w/n a feature list
void loopFeatureList(HASH * h, FILE * fp)
{
    struct stat fattr;
    static size_t optimal_size;
    ssize_t nr;
    static char *buf = NULL;

    //Determine optimal buffer size for file
    if (fstat(fileno(fp), &fattr))
	fatal_msg("fstat error.\n");

    //Find optimal size for Disk IO 
    optimal_size = (fattr.st_blksize >= BUFSIZ) ? fattr.st_blksize : BUFSIZ;

    buf = (char *) malloc(optimal_size);


    // replace w/ function pointers
    while ((nr = fread(buf, sizeof(char), optimal_size, fp)) != -1 && nr != 0) {

	    chkpushtxt(h, buf, nr);
    }

    printFeatures(h);
    freeHash(h);
}



// For finding all features

void loopRaw(HASH * h, FILE * fp)
{
    struct stat fattr;
    static size_t optimal_size;
    ssize_t nr;
    static char *buf = NULL;

    //Determine optimal buffer size for file
    if (fstat(fileno(fp), &fattr))
	fatal_msg("fstat error.\n");

    //Find optimal size for Disk IO 
    optimal_size = (fattr.st_blksize >= BUFSIZ) ? fattr.st_blksize : BUFSIZ;

    buf = (char *) malloc(optimal_size);


    // replace w/ function pointers
    while ((nr = fread(buf, sizeof(char), optimal_size, fp)) != -1 && nr != 0) {

	    pushtxt(h, buf, nr);
    }

    printFeatures(h);
    freeHash(h);
}
